cmd/hub/mcscontroller/controller.go

```diff
@@ -118,15 +121,205 @@ func (r *ServiceExportReconciler) Reconcile(ctx context.Context, req ctrl.Reques
+			//r.mcsClient.UpdateStatus(ctx, &se, metav1.UpdateOptions{})
+			err = r.deleteImport(ctx, &se)
+			if err != nil {
+				return ctrl.Result{}, err
+			}
+		} else { //3c:
+			r.updateServiceExportConditions(ctx, &se, mcsv1a1.ServiceExportValid, corev1.ConditionTrue, "", "")
+			r.Client.Status().Update(ctx, &se)
+			//se, err := r.mcsClient.UpdateStatus(ctx, &se, metav1.UpdateOptions{})
+			if err != nil {
+				return ctrl.Result{}, err
+			}
+			r.ensureImportFor(ctx, &se)
+		}
+	}
 	return ctrl.Result{}, nil
 }
 
+// getPrimaryExportObject finds the Primary ExportSpec from ServiceExportList.
+// Retruns the ExportSpec object that determined as the primary one from
+// the given ServiceExportList according to the conflict resolution specification set in the
+// KEP, that is implemented in IsPreferredOver method of ExportSpec.
+func getPrimaryExportObject(exportList mcsv1a1.ServiceExportList) (*mcs.ExportSpec, error) {
+	var primary *mcs.ExportSpec
+	for i, se := range exportList.Items {
+		exportSpec := &mcs.ExportSpec{}
+		err := exportSpec.UnmarshalObjectMeta(&se.ObjectMeta)
+		if err != nil {
+			return nil, err
+		}
+		if i == 0 || exportSpec.IsPreferredOver(primary) {
+			primary = exportSpec
+		}
+	}
+	return primary, nil
+}
+
+// create or update the ServiceImport corresponding to the provided export.
+func (r *ServiceExportReconciler) ensureImportFor(ctx context.Context, se *mcsv1a1.ServiceExport) error {
+	es := &mcs.ExportSpec{}
+
+	err := es.UnmarshalObjectMeta(&se.ObjectMeta)
+	if err != nil {
+		return err
+	}
+	namespacedName := types.NamespacedName{
+		Namespace: es.Name,
+		Name:      es.Namespace,
+	}
+	log := r.Log.WithValues("service", namespacedName.Name)
+
+	si := mcsv1a1.ServiceImport{}
+	err = r.Client.Get(ctx, namespacedName, &si)
+	expected := mcsv1a1.ServiceImport{
+		ObjectMeta: metav1.ObjectMeta{
+			Namespace: es.Namespace,
+			Name:      es.Name,
+		},
+		Spec: mcsv1a1.ServiceImportSpec{
+			Type:  es.Service.Type,
+			Ports: es.Service.Ports,
+		},
+	}
+	if apierrors.IsNotFound(err) {
+		err = r.Client.Create(ctx, &expected)
+		if err != nil {
+			log.Error(err, "unable to create the needed Service Import")
+			return err
+		}
+	} else if err == nil {
+		err = r.updateSi(ctx, &si, &expected, log)
+		if err != nil {
+			return err
+		}
+	} else {
+		log.Error(err, "Failed to get the needed service import")
+		return err
+	}
+	return nil
+}
+
+func (r *ServiceExportReconciler) updateSi(ctx context.Context, si *mcsv1a1.ServiceImport, expected *mcsv1a1.ServiceImport, log logr.Logger) error {
+	//not sure it's really needed to check name and namespace - I got the si from the Get according to those fields
+	//however its a general func, maybe should be "safe" for other usages(?)
+	if si.Name == expected.Name && si.Namespace == expected.Namespace {
+		si.Spec.Type = expected.Spec.Type
+		//maybe add a function that change the ports more carefully, without overwriting
+		si.Spec.Ports = expected.Spec.Ports
```

Comment:
The si.Spec.Type contains only some kind of enum/const string, so its safe just to override it.
However, the port field is more complicated. contains name, protocol, appProtocol, and port. as I understand, all those fields are critical, and if they differ in something, we will want to override them.

Reply:
you want to compare that the SI from Get matches the expected and only call update when they differ
you can use the DeepCopy function to copy all or set each field individually. Not sure what makes more sense (DeepCopy might overwrite some fields so probably field by field...)

Reply:
But if I'm anyway assigning the updated value to each port (or even spec) field, isn't it just implementing the deepCopy function? which field can be different and not be overwritten in this way? (if I'm assigning each field individually, but all fields).